input/script_utils.go

```diff
@@ -715,6 +1019,274 @@ func ReceiverHtlcSpendTimeout(signer Signer, signDesc *SignDescriptor,
+	htlcKey := txscript.ComputeTaprootOutputKey(
+		revokeKey, tapScriptRoot[:],
+	)
+
+	return &HtlcScriptTree{
+		TaprootKey:     htlcKey,
+		SuccessTapLeaf: successTapLeaf,
+		TimeoutTapLeaf: timeoutTapLeaf,
+		TapscriptTree:  tapscriptTree,
+		TapscriptRoot:  tapScriptRoot[:],
+	}, nil
+}
+
+// ReceiverHTLCScriptTaproot constructs the taproot witness program (schnor
+// key) for an outgoing HTLC on the receiver's version of the commitment
+// transaction. This method returns the top level tweaked public key that
+// commits to both the script paths. From the PoV fo teh receiver, this is an
+// accepted HTLC.
+//
+// The returned key commits to a tapscript tree with two possible paths:
+//
+//   - The timeout path:
+//     <remote_htlcpubkey> OP_CHECKSIG
+//     OP_CHECKSEQUENCEVERIFY
+//     <cltv_expiry> OP_CHECKLOCKTIMEVERIFY OP_DROP
+//
+//   - Success path:
+//     OP_SIZE 32 OP_EQUALVERIFY
+//     OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY
+//     <local_htlcpubkey> OP_CHECKSIGVERIFY
+//     <remote_htlcpubkey> OP_CHECKSIG
+//
+// The timeout path can be be spent with a witness of:
+//   - <sender sig> <timeout_script> <control_block>
+//
+// The success path can be spent with a witness of:
+//   - <sender sig> <receiver sig> <preimage> <success_script> <control_block>
+//
+// The top level keyspend key is the revocation key, which allows a defender to
+// unilaterally spend the created output. Both the final output key as well as
+// the tap leaf are returned.
+func ReceiverHTLCScriptTaproot(cltvExpiry uint32,
+	senderHtlcKey, receiverHtlcKey, revocationKey *btcec.PublicKey,
+	payHash []byte) (*HtlcScriptTree, error) {
+
+	// Given all the necessary parameters, we'll return the HTLC script
+	// tree that includes the top level output script, as well as the two
+	// tap leaf paths.
+	return receiverHtlcTapScriptTree(
+		senderHtlcKey, receiverHtlcKey, revocationKey, payHash,
+		cltvExpiry,
+	)
+}
+
+// ReceiverHTLCScriptTaprootRedeem creates a valid witness needed to redeem a
+// receiver taproot HTLC with the pre-image. The returned witness is valid and
+// includes the control block required to spend the output.
+func ReceiverHTLCScriptTaprootRedeem(senderSig Signature,
+	senderSigHash txscript.SigHashType, paymentPreimage []byte,
+	signer Signer, signDesc *SignDescriptor,
+	htlcSuccessTx *wire.MsgTx, revokeKey *btcec.PublicKey,
+	tapscriptTree *txscript.IndexedTapScriptTree) (wire.TxWitness, error) {
+
+	// First, we'll generate a signature for the HTLC success transaction.
+	// The signDesc should be signing with the public key used as the
+	// receiver's public key and also the correct single tweak.
+	sweepSig, err := signer.SignOutputRaw(htlcSuccessTx, signDesc)
+	if err != nil {
+		return nil, err
+	}
+
+	// In addition to the signature and the witness/leaf script, we also
+	// need to make a control block proof using the tapscript tree.
+	timeoutTapLeafHash := txscript.NewBaseTapLeaf(
```

Comment:
should be `successLeafHash`

Reply:
Lines have changed, but i think current diff is fine?